python/bap.py

```diff
@@ -0,0 +1,342 @@
+            return self.symbols
+        elif name == 'addr' or name == 'size':
+            return self.get('memory')[name]
+        elif name == 'memory':
+            self.memory = Memory(self.get('memory'), self)
+            return self.memory
+        else:
+            return self.get(name)
+
+class Symbol(Resource):
+    def __init__(self, ident, parent):
+        super(Symbol, self).__init__('symbol', ident, parent.bap)
+        self.parent = parent
+
+    def load_chunks(self):
+        self.chunks = [Memory(s, self) for s in self.get('chunks')]
+
+    def __getattr__(self, name):
+        if name == 'chunks':
+            self.load_chunks()
+            return self.chunks
+        elif name == 'addr':
+            self.load_chunks()
+            return self.chunks[0].addr
+        else:
+            return self.get(name)
+
+class Memory(object):
+    def __init__(self, mem, parent):
+        self.parent = parent
+        self.size = int(mem['size'])
+        self.addr = int(mem['addr'])
+        self.links = mem['links']
+
+    def load_data(self):
+        try:
+            url = (urlparse(url) for url in self.links
+                   if urlparse(url).scheme == 'mmap').next()
+            qs = parse_qs(url.query)
+            offset = int(qs['offset'][0])
+            with open(url.path, "rw+b") as f:
+                mm = mmap(f.fileno(), length=0)
+                mm.seek(offset)
+                self.data = mm.read(self.size)
+                mm.close()
+        except StopIteration:
+            self.data = None
+
+    def __getattr__(self, name):
+        if name == 'data':
+            self.load_data()
+            return self.data
+        raise AttributeError(name)
+
+
+class ServerError(Exception):
+    def __init__(self, err):
+        self.msg = str(Error(err))
+
+    def __str__(self):
+        return self.msg
+
+class Error(object):
+    def __init__(self, err):
+        self.__dict__.update(err)
+        self.__dict__.update(err['error'])
+
+    def __str__(self):
+        return "{severity}: {description}".format(**self.error)
+
+class Id(object):
+    def __init__(self, r):
+        self.value = r
+    def __str__(self):
+        return str(self.value)
+
+RETRIES = 10
+
+class Bap(object):
+    def __init__(self, server={}):
+        if isinstance(server, dict):
+            self.__dict__.update(spawn_server(**server))
+        else:
+            self.url = server
+
+        self.last_id = 0
+        for attempt in range(RETRIES):
+            try:
+                self.capabilities = self.call({'init' : {
+                    'version' : '0.1'}})
+            except Exception:
+                if attempt + 1 == RETRIES:
+                    raise
+                else:
+                    time.sleep(0.1 * attempt)
+
+        if not "capabilities" in self.__dict__:
+            raise RuntimeError("Failed to connect to BAP server")
+
+        self.data = {}
+        self.temp = NamedTemporaryFile('rw+b')
```

Comment:
If we support shm as was stated in the README, and we're spawning on the same server, why are we not using it?